expression/util.go

```diff
@@ -526,3 +527,44 @@ func GetParamExpression(ctx sessionctx.Context, v *driver.ParamMarkerExpr, useCa
 	}
 	return value, nil
 }
+
+// ParamToByItemNode generate ByItem node from ParamMarkerExpr.
+func ParamToByItemNode(ctx sessionctx.Context, v *driver.ParamMarkerExpr) (*ast.ByItem, bool, error) {
+	value, err := GetParamExpression(ctx, v)
+	if err != nil {
+		return nil, true, errors.Trace(err)
+	}
+	pos, isNull, err := GetIntFromConstant(ctx, value)
+	if err != nil || isNull {
+		return nil, true, errors.Trace(err)
+	}
+	byItem := &ast.ByItem{Expr: &ast.PositionExpr{N: pos, P: v}}
```

Comment:
Yes, we should because we don't cache the plan having the parameterized ``ORDER BY`` clause.

Reply:
I come up with another idea. How about don't calculate `N`, just construct a deferred function in `P`? Thus we don't need to modify this function when we support plan cache for these statements?

Reply:
How about to handle it as another PR?

Reply:
Unfortunately, we can't now. Sorry for the confusion. To produce the same error message like MySQL, we should know that a parameterized position is out of range. ``P`` indicates that a position is parameterized. Here are the test case:

https://github.com/pingcap/tidb/blob/fe31c7405df2df85fa9d9aba7037090379242341/executor/prepared_test.go#L728-L730

Reply:
If so, calculating `N` in here is not needed, because the calculated value is never used. Can we remove this calculation process? i.e. remove the function call of `GetParamExpression()` and `GetIntFromConstant()`.

Reply:
@eurekaka 
``P`` can be a ``ValueExpr`` or ``Constant`` expression because its type is ``ExprNode``.  ``UseCache`` only indicates that ``P`` can be deferred constant or not. That is, we can't decide that ``P`` is set to ``nil`` depending on ``UseCache``.